controller/internal/enforcer/dnsproxy/dns.go

```diff
@@ -0,0 +1,225 @@
+	var ips []string
+	c := new(dns.Client)
+	c.Dialer = &net.Dialer{
+		Control: func(_, _ string, c syscall.RawConn) error {
+			return c.Control(func(fd uintptr) {
+				if err := syscall.SetsockoptInt(int(fd), syscall.SOL_SOCKET, syscall.SO_MARK, proxyMarkInt); err != nil {
+					zap.L().Error("Failed to assing mark to socket", zap.Error(err))
+				}
+			})
+		},
+		Timeout: dnsRequestTimeout,
+	}
+
+	in, _, err := c.Exchange(r, net.JoinHostPort(ip.String(), strconv.Itoa(int(port))))
+	if err != nil {
+		return nil, nil, 0, err
+	}
+
+	var ttl uint32
+
+	for _, ans := range in.Answer {
+		if ans.Header().Rrtype == dns.TypeA {
+			t, _ := ans.(*dns.A)
+			ips = append(ips, t.A.String())
+			if ans.Header().Ttl > ttl {
+				ttl = ans.Header().Ttl
+			}
+		}
+
+		if ans.Header().Rrtype == dns.TypeAAAA {
+			t, _ := ans.(*dns.AAAA)
+			ips = append(ips, t.AAAA.String())
+			if ans.Header().Ttl > ttl {
+				ttl = ans.Header().Ttl
+			}
+		}
+	}
+
+	return in, ips, ttl, nil
+}
+
+func (s *serveDNS) ServeDNS(w dns.ResponseWriter, r *dns.Msg) {
+	var err error
+	lAddr := w.LocalAddr().(*net.UDPAddr)
+	rAddr := w.RemoteAddr().(*net.UDPAddr)
+	var puCtx *pucontext.PUContext
+
+	defer func() {
+		if puCtx != nil {
+			var errStr string
+			if err != nil {
+				errStr = err.Error()
+			}
+			s.reportDNSLookup(r.Question[0].Name, puCtx, rAddr.IP, errStr)
+		}
+	}()
+
+	origIP, origPort, _, err := s.conntrack.GetOriginalDest(net.ParseIP("127.0.0.1"), rAddr.IP, uint16(lAddr.Port), uint16(rAddr.Port), 17)
+	if err != nil {
+		zap.L().Error("Failed to find flow for the redirected dns traffic", zap.Error(err))
+		return
+	}
+
+	data, err := s.puFromID.Get(s.contextID)
+	if err != nil {
+		zap.L().Error("context not found for the PU with ID", zap.String("contextID", s.contextID))
+		return
+	}
+
+	dnsReply, ips, ttl, err := forwardDNSReq(r, origIP, origPort)
+	if err != nil {
+		zap.L().Debug("Forwarded dns request returned error", zap.Error(err))
+		return
+	}
+
+	puCtx = data.(*pucontext.PUContext)
+	ps, err1 := puCtx.GetPolicyFromFQDN(r.Question[0].Name)
+	if err1 == nil {
+		for _, p := range ps {
+			if err1 := puCtx.UpdateApplicationACLs(policy.IPRuleList{{Addresses: ips,
+				Ports:     p.Ports,
+				Protocols: p.Protocols,
+				Policy:    p.Policy,
+			}}); err1 != nil {
+				zap.L().Error("Adding IP rule returned error", zap.Error(err1))
+			}
+		}
+	}
+
+	if err = w.WriteMsg(dnsReply); err != nil {
+		zap.L().Error("Writing dns response back to the client returned error", zap.Error(err))
+	}
+
+	go func() {
+		time.Sleep(time.Duration(ttl) * time.Second)
+
+		data, err := s.puFromID.Get(s.contextID)
+		if err != nil {
+			return
+		}
+
+		puCtx := data.(*pucontext.PUContext)
+
+		for _, stringip := range ips {
+			ip := net.ParseIP(stringip)
+			var mask int
+			if ip.To4() == nil {
+				mask = 128
+			} else {
+				mask = 32
+			}
+			puCtx.RemoveApplicationACL(ip, mask)
+		}
+	}()
```

Comment:
I am little concerned about that. This will start a go-routine for every DNS request essentially. This can easily become a vector for an attack. Can't we just poll the table periodically and avoid per-request go-routines?